File: intellij-plugin/core/src/test/kotlin/com/decoy/core/GatewayConnectivityTest.kt
package com.decoy.core

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

private val CREDS = GatewayCheckCredentials(authToken = "fake-token", baseUrl = "https://gw.example.com", skipTlsVerify = false)

class GatewayConnectivityTest {

    @Test
    fun `ok - 200 response classifies as ok, with the correct Bearer header and URL sent`() {
        val calls = mutableListOf<Triple<String, Map<String, String>, Boolean>>()
        val requestFn = GatewayRequestFn { url, headers, _, skipTlsVerify ->
            calls.add(Triple(url, headers, skipTlsVerify))
            GatewayHttpResponse(200, "{}")
        }
        val result = checkGatewayConnectivity(CREDS, requestFn)
        assertEquals(GatewayCheckResult.Ok, result)
        assertEquals("https://gw.example.com/v1/messages", calls[0].first)
        assertEquals("Bearer fake-token", calls[0].second["authorization"])
        assertFalse(calls[0].third)
    }

    @Test
    fun `401 classifies as unauthorized with a message naming the fix`() {
        val requestFn = GatewayRequestFn { _, _, _, _ -> GatewayHttpResponse(401, "unauthorized") }
        val result = checkGatewayConnectivity(CREDS, requestFn)
        assertTrue(result is GatewayCheckResult.Failed)
        result as GatewayCheckResult.Failed
        assertEquals(GatewayCheckResult.Failed.Kind.UNAUTHORIZED, result.kind)
        assertTrue(result.message.contains("token", ignoreCase = true))
    }

    @Test
    fun `403 also classifies as unauthorized`() {
        val requestFn = GatewayRequestFn { _, _, _, _ -> GatewayHttpResponse(403, "forbidden") }
        val result = checkGatewayConnectivity(CREDS, requestFn) as GatewayCheckResult.Failed
        assertEquals(GatewayCheckResult.Failed.Kind.UNAUTHORIZED, result.kind)
    }

    @Test
    fun `a thrown exception (unreachable host) classifies as unreachable, not a crash`() {
        val requestFn = GatewayRequestFn { _, _, _, _ -> throw java.net.UnknownHostException("gw.example.com") }
        val result = checkGatewayConnectivity(CREDS, requestFn)
        assertTrue(result is GatewayCheckResult.Failed)
        result as GatewayCheckResult.Failed
        assertEquals(GatewayCheckResult.Failed.Kind.UNREACHABLE, result.kind)
        assertTrue(result.message.contains("gw.example.com"))
    }

    @Test
    fun `a non-401,403 4xx,5xx classifies as http_error, body included for diagnosis`() {
        val requestFn = GatewayRequestFn { _, _, _, _ -> GatewayHttpResponse(500, "internal gateway error") }
        val result = checkGatewayConnectivity(CREDS, requestFn) as GatewayCheckResult.Failed
        assertEquals(GatewayCheckResult.Failed.Kind.HTTP_ERROR, result.kind)
        assertTrue(result.message.contains("500"))
        assertTrue(result.message.contains("internal gateway error"))
    }

    @Test
    fun `trailing slash on baseUrl does not produce a double slash in the request URL`() {
        val calls = mutableListOf<String>()
        val requestFn = GatewayRequestFn { url, _, _, _ ->
            calls.add(url)
            GatewayHttpResponse(200, "{}")
        }
        checkGatewayConnectivity(CREDS.copy(baseUrl = "https://gw.example.com/"), requestFn)
        assertEquals("https://gw.example.com/v1/messages", calls[0])
    }
}

File: vscode-extension/src/gatewayConnectivity.test.ts
import { checkGatewayConnectivity, GatewayRequestFn } from "./gatewayConnectivity";

const CREDS = { authToken: "fake-token", baseUrl: "https://gw.example.com", skipTlsVerify: false };

describe("checkGatewayConnectivity", () => {
  it("ok: 200 response classifies as ok, with the correct Bearer header and URL sent", async () => {
    const calls: Array<{ url: string; headers: Record<string, string>; skipTlsVerify: boolean }> = [];
    const requestFn: GatewayRequestFn = async (url, headers, _body, skipTlsVerify) => {
      calls.push({ url, headers, skipTlsVerify });
      return { status: 200, body: "{}" };
    };
    const result = await checkGatewayConnectivity(CREDS, requestFn);
    expect(result).toEqual({ ok: true });
    expect(calls[0].url).toBe("https://gw.example.com/v1/messages");
    expect(calls[0].headers.authorization).toBe("Bearer fake-token");
    expect(calls[0].skipTlsVerify).toBe(false);
  });

  it("401 classifies as unauthorized with a message naming the fix", async () => {
    const requestFn: GatewayRequestFn = async () => ({ status: 401, body: "unauthorized" });
    const result = await checkGatewayConnectivity(CREDS, requestFn);
    expect(result.ok).toBe(false);
    if (!result.ok) {
      expect(result.kind).toBe("unauthorized");
      expect(result.message).toMatch(/token/i);
    }
  });

  it("403 also classifies as unauthorized", async () => {
    const requestFn: GatewayRequestFn = async () => ({ status: 403, body: "forbidden" });
    const result = await checkGatewayConnectivity(CREDS, requestFn);
    expect(result.ok).toBe(false);
    if (!result.ok) expect(result.kind).toBe("unauthorized");
  });

  it("a thrown network error (unreachable host) classifies as unreachable, not a crash", async () => {
    const requestFn: GatewayRequestFn = async () => {
      throw new Error("getaddrinfo ENOTFOUND gw.example.com");
    };
    const result = await checkGatewayConnectivity(CREDS, requestFn);
    expect(result.ok).toBe(false);
    if (!result.ok) {
      expect(result.kind).toBe("unreachable");
      expect(result.message).toContain("gw.example.com");
    }
  });

  it("a non-401/403 4xx/5xx classifies as http_error, body included for diagnosis", async () => {
    const requestFn: GatewayRequestFn = async () => ({ status: 500, body: "internal gateway error" });
    const result = await checkGatewayConnectivity(CREDS, requestFn);
    expect(result.ok).toBe(false);
    if (!result.ok) {
      expect(result.kind).toBe("http_error");
      expect(result.message).toContain("500");
      expect(result.message).toContain("internal gateway error");
    }
  });

  it("trailing slash on baseUrl does not produce a double slash in the request URL", async () => {
    const calls: string[] = [];
    const requestFn: GatewayRequestFn = async (url) => {
      calls.push(url);
      return { status: 200, body: "{}" };
    };
    await checkGatewayConnectivity({ ...CREDS, baseUrl: "https://gw.example.com/" }, requestFn);
    expect(calls[0]).toBe("https://gw.example.com/v1/messages");
  });
});
